platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/DecoratedStackImpl.java

```diff
@@ -0,0 +1,335 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2016 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.workspace.server.model.impl.stack;
+
+import org.eclipse.che.api.core.model.workspace.WorkspaceConfig;
+import org.eclipse.che.api.machine.shared.Permissions;
+import org.eclipse.che.api.workspace.server.stack.image.StackIcon;
+import org.eclipse.che.commons.lang.NameGenerator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Server implementation of {@link Stack}
+ *
+ * @author Alexander Andrienko
+ */
+public class DecoratedStackImpl implements DecoratedStack {
+
+    public static StackBuilder builder() {
+        return new StackBuilder();
+    }
+
+    public DecoratedStackImpl(Stack stack, StackIcon stackIcon) {
+        this(stack.getId(),
+             stack.getName(),
+             stack.getDescription(),
+             stack.getScope(),
+             stack.getCreator(),
+             stack.getTags(),
+             stack.getWorkspaceConfig(),
+             stack.getSource(),
+             stack.getComponents(),
+             stackIcon,
+             stack.getPermissions());
+    }
+
+    public DecoratedStackImpl(String id,
+                              String name,
+                              String description,
+                              String scope,
+                              String creator,
+                              List<String> tags,
+                              WorkspaceConfig workspaceConfig,
+                              StackSource source,
+                              List<? extends StackComponent> components,
+                              StackIcon icon,
+                              Permissions permissions) {
+        this.id = id;
```

Comment:
Stack should not exist without the id, [example](https://github.com/codenvy/che-core/blob/2f6667b4a147ea1c229722b74c98f2b5c346e27b/platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/impl/SnapshotImpl.java#L68-L68)